src/python.cc

```diff
@@ -2121,6 +2325,461 @@ ModelState::ModelState(TRITONBACKEND_Model* triton_model)
+          TRITONSERVER_ERROR_INTERNAL,
+          (std::string(
+               "Received unexpected resposne from Python backend stub: ") +
+           name_)
+              .c_str());
+    }
+
+    auto auto_complete_response =
+        std::move((shm_pool_->Load<AutoCompleteResponseShm>(
+                      auto_complete_response_message->Args())))
+            .data_;
+
+    if (auto_complete_response->response_has_error) {
+      if (auto_complete_response->response_is_error_set) {
+        std::unique_ptr<PbString> error_message =
+            PbString::LoadFromSharedMemory(
+                shm_pool_, auto_complete_response->response_error);
+        return TRITONSERVER_ErrorNew(
+            TRITONSERVER_ERROR_INTERNAL, error_message->String().c_str());
+      } else {
+        return TRITONSERVER_ErrorNew(
+            TRITONSERVER_ERROR_INTERNAL,
+            (std::string("Auto-complete failed for ") + Name()).c_str());
+      }
+    }
+
+    if (auto_complete_response->response_has_model_config) {
+      std::unique_ptr<PbString> auto_complete_config =
+          PbString::LoadFromSharedMemory(
+              shm_pool_, auto_complete_response->response_model_config);
+      std::string auto_complete_config_string = auto_complete_config->String();
+      // Need to corret the string for the right json format
+      CorrectStringFormat(&auto_complete_config_string);
+      RETURN_IF_ERROR(auto_complete_config_.Parse(auto_complete_config_string));
+      triton::common::TritonJson::WriteBuffer buf;
+      auto_complete_config_.Write(&buf);
+    }
+
+    initialized_ = true;
+  }
+
+  return nullptr;  // success
+}
+
+void
+ModelState::CorrectStringFormat(std::string* str)
+{
+  *str = std::regex_replace(*str, std::regex("\'"), "\"");
+  *str = std::regex_replace(*str, std::regex("False"), "false");
+  *str = std::regex_replace(*str, std::regex("True"), "true");
+}
+
+TRITONSERVER_Error*
+ModelState::AutoCompleteConfig()
+{
+  RETURN_IF_ERROR(CheckModelConfigMismatch());
+
+  // If the model configuration already specifies inputs and outputs
+  // then don't perform any auto-completion.
+  size_t input_cnt = 0;
+  size_t output_cnt = 0;
+  {
+    triton::common::TritonJson::Value inputs;
+    if (ModelConfig().Find("input", &inputs)) {
+      input_cnt = inputs.ArraySize();
+    }
+
+    triton::common::TritonJson::Value config_batch_inputs;
+    if (ModelConfig().Find("batch_input", &config_batch_inputs)) {
+      input_cnt += config_batch_inputs.ArraySize();
+    }
+
+    triton::common::TritonJson::Value outputs;
+    if (ModelConfig().Find("output", &outputs)) {
+      output_cnt = outputs.ArraySize();
+    }
+  }
+
+  // Determine if the model can potentially support batching. All
+  // input and output tensors must have a variable first dimension.
+  // However, ragged batching is an exception to this rule. A tensor
+  // allowing ragged batch is in itself a batching hint.
+  bool config_batch_hint = false;
```

Comment:
Much of the logic for checking batching support can be be simplified. TF/ONNX/TRT models provide the exact shape that is expected by the model. For example take a model with input and shape: INPUT0 : [-1, 16]. There is no way to detect whether to enable batching on these model while auto-completing. Hence, we look for the hint in config.pbtxt file. If in the file they have provided INPUT0 as [16]. Then that means the batching needs to be enabled.

For us with the current design it is simpler. The python model if it supports batching will call set_max_batch_size with a non-zero value in the model.py. It will still set INPUT0 as [16] as we are providing the config dim in the model.py. Because the dim we are providing is config dim and not model dim, we can not actually look for such hints here. You will see it yourself if you try the batch_model.py, I proposed above.

The logic here should be simple. If the user has provided non-zero max_batch_size in model.py or in model.pbtxt, then treat it as a batching model. If both are non-zero and there is a mismatch between the two sizes then throw an error.

Reply:
Thanks for the clarification and the link. The logic is updated. Please review.